### PR TITLE
feat: bucket

### PR DESCRIPTION
**Pull request by thantos**
* [x] get, put, delete, list, stream, presigned url
* [x] Call from handlers
* [x] Call from workflow
* [x] Infer
* [x] Test

### CODE REVIEW

File: apps/tests/aws-runtime/test/test-service.ts [hunk 1]
```diff
@@ -672,6 +674,87 @@ export const transactionWorkflow = workflow(
   }
 );
 
+export const myBucket = bucket("myBucket");
+export const bucketSignal = signal<{ data: string }>("bucketSignal");
+
+export const myBucketStream = myBucket.stream(
```
Comment (sam-goodwin) [inline apps/tests/aws-runtime/test/test-service.ts:680]: is `stream` the right word here?

File: packages/@eventual/aws-cdk/src/bucket-service.ts [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                allowedHeaders: props.serviceProps.cors.allowHeaders,
+                exposedHeaders: props.serviceProps.cors.exposeHeaders,
+                maxAge: props.serviceProps.cors.maxAge?.toSeconds(),
+              },
+            ]
+          : undefined,
+      bucketName: bucketServiceBucketName(
+        props.serviceProps.serviceName,
+        props.bucket.name
+      ),
```
Comment (sam-goodwin) [inline packages/@eventual/aws-cdk/src/bucket-service.ts:161]: bucket names are the most problematic because of global uniqueness. Do you have any thoughts on that?

File: packages/@eventual/aws-cdk/src/service.ts [hunk 3]
```diff
@@ -176,10 +183,21 @@ export class Service<S = any> extends Construct {
...
+   */
   public readonly entityStreams: ServiceEntityStreams<S>;
+  /**
+   * Buckets defined by the service.
+   */
+  public readonly buckets: ServiceBuckets<S>;
+  /**
+   * Streams of bucket events defined by the service.
+   */
+  public readonly bucketSteams: ServiceBucketStreams<S>;
```
Comment (sam-goodwin) [inline packages/@eventual/aws-cdk/src/service.ts:200]: typo

File: packages/@eventual/aws-runtime/src/handlers/bucket-stream-worker.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  );
+
+  if (results.rejected.length > 0) {
+    console.log(
+      "Events failed \n",
+      results.rejected
+        .map(([record, error]) => `${record.s3.object.key} - ${error}`)
+        .join("\n")
+    );
+  }
```
Comment (sam-goodwin) [inline packages/@eventual/aws-runtime/src/handlers/bucket-stream-worker.ts:56]: does the contract support returning failed records? I.e. "partial sucess"

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+          eTag: c.ETag!,
+          key: c.Key!,
+          size: c.Size ?? 0,
+        })) ?? [],
+    };
+  }
+
+  private bucketName(bucketName: string) {
+    return bucketServiceBucketName(getLazy(this.props.serviceName), bucketName);
+  }
```
Comment (sam-goodwin) [inline packages/@eventual/aws-runtime/src/stores/bucket-store.ts:174]: Should this be a public getter?

File: packages/@eventual/core-runtime/src/local/stores/bucket-store.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+  > = {};
+
+  // TODO: publish events for the stream
+  constructor(private props: LocalBucketStoreProps) {}
+  async get(
+    bucketName: string,
+    key: string,
+    options?: GetBucketObjectOptions
+  ): Promise<GetBucketObjectResponse | undefined> {
+    const bucket = this.objects[bucketName];
```
Comment (sam-goodwin) [inline packages/@eventual/core-runtime/src/local/stores/bucket-store.ts:42]: i assume we may want to store these on the file system one day

File: packages/@eventual/core-runtime/src/local/stores/bucket-store.ts [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    };
+  }
+
+  presignedUrl(
+    _bucketName: string,
+    _key: string,
+    _operation: "get" | "put" | "delete",
+    _expires?: DurationSchedule | undefined
+  ): Promise<{ url: string; expires: string }> {
+    throw new Error("Presigned urls are not supported in Eventual Local");
```
Comment (sam-goodwin) [inline packages/@eventual/core-runtime/src/local/stores/bucket-store.ts:163]: What would it take to support this?

File: packages/@eventual/core-runtime/src/utils.ts [hunk 8]
```diff
@@ -64,3 +69,48 @@ export function deserializeCompositeKey(
...
+    (!streamSpec.options?.namespaces ||
+      (item.namespace &&
+        streamSpec.options.namespaces.includes(item.namespace))) &&
+    (!streamSpec.options?.namespacePrefixes ||
+      (item.namespace &&
+        streamSpec.options.namespacePrefixes.some((p) =>
+          item.namespace?.startsWith(p)
+        )))
+  );
+}
```
Comment (sam-goodwin) [inline packages/@eventual/core-runtime/src/utils.ts:98]: What is this doing?

File: packages/@eventual/core-runtime/src/utils.ts [hunk 9]
```diff
@@ -64,3 +69,48 @@ export function deserializeCompositeKey(
...
+      streamSpec.options?.operations.includes(item.operation)) &&
+    (!streamSpec.options?.filters ||
+      streamSpec.options?.filters.some((f) => {
+        return (
+          (!f.prefix || item.key.startsWith(f.prefix)) &&
+          (!f.suffix || item.key.endsWith(f.suffix))
+        );
+      }))
+  );
+}
```
Comment (sam-goodwin) [inline packages/@eventual/core-runtime/src/utils.ts:116]: What is this doing?

File: packages/@eventual/core/src/bucket.ts [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+  isSourceLocation,
+  SourceLocation,
+} from "./internal/service-spec.js";
+import { DurationSchedule } from "./schedule.js";
+
+export interface BucketObjectReference {
+  key: string;
+  eTag: string;
+  size: number;
+}
```
Comment (sam-goodwin) [inline packages/@eventual/core/src/bucket.ts:17]: Might be worth cross-referencing these types with CloudFlare's R2 https://developers.cloudflare.com/r2/api/workers/workers-api-reference/#method-specific-types

**Review by sam-goodwin — CHANGES_REQUESTED** [review]

File: apps/tests/aws-runtime/test/test-service.ts (hunk 1 above)
Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts:680]: Nope, but I picked a word until we could debate it.

Notify? (They call them notifications)
OnNotification
OnChange
Changes
OnPut/OnDelete

File: packages/@eventual/aws-cdk/src/bucket-service.ts (hunk 2 above)
Comment (thantos) [reply packages/@eventual/aws-cdk/src/bucket-service.ts:161]: Just that the service name would make it more unique

Another idea was to use a single bucket and split internally, but that could create issues if trying to make a static site or cdn.

Or add a unique value to the name, which is already limited length.

File: packages/@eventual/aws-cdk/src/service.ts (hunk 3 above)
Comment (thantos) [reply packages/@eventual/aws-cdk/src/service.ts:200]: Ha

File: packages/@eventual/aws-runtime/src/handlers/bucket-stream-worker.ts (hunk 4 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/handlers/bucket-stream-worker.ts:56]: No. They are async only. No retry, no success/failure.

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 5 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts:174]: It's just a local helper, any use of bucket name is internal.

File: packages/@eventual/core-runtime/src/local/stores/bucket-store.ts (hunk 6 above)
Comment (thantos) [reply packages/@eventual/core-runtime/src/local/stores/bucket-store.ts:42]: Likely. There is a larger story to persist local data across restarts.

File: packages/@eventual/core-runtime/src/local/stores/bucket-store.ts (hunk 7 above)
Comment (thantos) [reply packages/@eventual/core-runtime/src/local/stores/bucket-store.ts:163]: I think we'd need a system path that supports the president contract.

File: packages/@eventual/core-runtime/src/utils.ts (hunk 9 above)
Comment (thantos) [reply packages/@eventual/core-runtime/src/utils.ts:116]: Supports filtering logic for local.

File: apps/tests/aws-runtime/test/test-service.ts (hunk 1 above)
Comment (sam-goodwin) [reply apps/tests/aws-runtime/test/test-service.ts:680]: Could do idiomatic node js APIs. 

```
bucket.on("put")
```

File: packages/@eventual/aws-cdk/src/bucket-service.ts (hunk 2 above)
Comment (sam-goodwin) [reply packages/@eventual/aws-cdk/src/bucket-service.ts:161]: Perhaps we just allow the name to be overridden. Problem solved.

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 5 above)
Comment (sam-goodwin) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts:174]: That makes escape hatches to aws s3 sdk impossible. Why not expose it?

File: packages/@eventual/core-runtime/src/local/stores/bucket-store.ts (hunk 7 above)
Comment (sam-goodwin) [reply packages/@eventual/core-runtime/src/local/stores/bucket-store.ts:163]: Let's track?

File: apps/tests/aws-runtime/test/test-service.ts (hunk 1 above)
Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts:680]: We could, though I'd like to be able to handle multiple event types from one handler, in an arbitrary set.

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 5 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts:174]: We don't expose the raw resource name for anything else. Won't it be confusing for a Bucket to have a name and a bucket name? 

We could create a property called `awsBucketName`. I assume the bucket name discussed above would be defined in the service side? 

If the override is defined in the CDK side, we'd need to rely on synthesis and maybe deployment to get the value (not the service spec like all other values). This value would not exist for local/testing as well as local doesn't have AWS. the `awsBucketName` would be undefned.

I'd argue lets follow the current pattern of hiding the aws layer until this value is needed.

File: packages/@eventual/core/src/bucket.ts (hunk 10 above)
Comment (thantos) [reply packages/@eventual/core/src/bucket.ts:17]: Looks fine, `etag` is `matchIf` in s3 and `onlyIf` in r2. The only difference is r2 doesn't return metadata about objects like size, which can be 0 for cloudflare.

File: packages/@eventual/core-runtime/src/utils.ts (hunk 8 above)
Comment (thantos) [reply packages/@eventual/core-runtime/src/utils.ts:98]: jsut moved this from core, implementation of the stream matching logic for use in code (ex: local).

File: packages/@eventual/core-runtime/src/local/stores/bucket-store.ts (hunk 7 above)
Comment (thantos) [reply packages/@eventual/core-runtime/src/local/stores/bucket-store.ts:163]: https://github.com/functionless/eventual/issues/341

File: apps/tests/aws-runtime/test/test-service.ts (hunk 1 above)
Comment (sam-goodwin) [reply apps/tests/aws-runtime/test/test-service.ts:680]: Can always pass an array .on(["put"])

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 5 above)
Comment (sam-goodwin) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts:174]: I don't think it's confusing. It's safe to assume a bucket in any system has a physical name.

File: apps/tests/aws-runtime/test/test-service.ts (hunk 1 above)
Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts:680]: Yeah, thats what I was thinking too.

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 5 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts:174]: Yeah, thats true, and the physical name can be different between envs.

File: apps/tests/aws-runtime/test/test-service.ts [hunk 11]
```diff
@@ -672,6 +675,109 @@ export const transactionWorkflow = workflow(
...
+async function streamToString(stream: stream.Readable) {
+  // lets have a ReadableStream as a stream variable
+  const chunks = [];
+
+  for await (const chunk of stream) {
+    chunks.push(Buffer.from(chunk));
+  }
+
+  return Buffer.concat(chunks).toString("utf-8");
+}
```
Comment (sam-goodwin) [inline apps/tests/aws-runtime/test/test-service.ts:706]: Is the returned object always a stream?

No way should we expect customers to always do this. The library should have two different methods, one for stream, one for string/Buffer?

File: packages/@eventual/aws-cdk/src/bucket-service.ts [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+  }
+}
+
+interface BucketProps {
+  serviceProps: BucketServiceProps<any>;
+  bucketService: BucketService<any>;
+  bucket: BucketRuntime;
+}
+
+export class Bucket extends Construct {
```
Comment (sam-goodwin) [inline packages/@eventual/aws-cdk/src/bucket-service.ts:159]: does this need to be exported?

File: packages/@eventual/aws-cdk/src/bucket-service.ts [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+                      (v) => s.toString() === v.toString()
+                    ) ?? []
+                ),
+                allowedOrigins: props.serviceProps.cors.allowOrigins,
+                allowedHeaders: props.serviceProps.cors.allowHeaders,
+                exposedHeaders: props.serviceProps.cors.exposeHeaders,
+                maxAge: props.serviceProps.cors.maxAge?.toSeconds(),
+              },
+            ]
+          : undefined,
```
Comment (sam-goodwin) [inline packages/@eventual/aws-cdk/src/bucket-service.ts:188]: Curious, why was this needed?

File: packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  entityClient: createEntityClient(),
+  serviceClient: createServiceClient({}),
+  serviceName,
+  serviceSpec,
+  serviceUrl,
+});
+
+export default (async (event) => {
+  const records = event.Records;
+  console.log("records", JSON.stringify(records, undefined, 4));
```
Comment (sam-goodwin) [inline packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts:34]: we really need debug logs ...

File: packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+  const results = await promiseAllSettledPartitioned(
+    records,
+    async (record) => {
+      await worker({
+        bucketName: getLazy(bucketName),
+        handlerName: getLazy(bucketHandlerName),
```
Comment (sam-goodwin) [inline packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts:41]: i forget, what does `getLazy` mean? the name doesn't make sense to me .. this is not a promise, what's lazy about it

File: packages/@eventual/aws-runtime/src/handlers/command-worker.ts [hunk 16]
```diff
@@ -19,8 +20,11 @@ import { serviceName } from "../env.js";
  */
 export default createApiGCommandAdaptor({
   commandWorker: createCommandWorker({
-    // the service client, spec, and service url will be created at runtime, using a computed uri from the apigateway request
+    bucketStore: createBucketStore(),
     entityClient: createEntityClient(),
+    // the service client, spec, and service url will be created at runtime, using a computed uri from the apigateway request
+    serviceClient: undefined,
+    serviceSpec: undefined,
```
Comment (sam-goodwin) [inline packages/@eventual/aws-runtime/src/handlers/command-worker.ts:27]: Why is that needed?

File: apps/tests/aws-runtime/test/test-service.ts (hunk 11 above)
Comment (sam-goodwin) [reply apps/tests/aws-runtime/test/test-service.ts:706]: We could add a method to the returned stream that stringifies it, `toString("utf8")`, `toBuffer()`. Is there any precedent for this?

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+        CopySourceIfMatch: options?.sourceEtag,
+      })
+    );
+
+    return {
+      etag: result.CopyObjectResult?.ETag,
+    };
+  }
+
+  public async presignedUrl(
```
Comment (sam-goodwin) [inline packages/@eventual/aws-runtime/src/stores/bucket-store.ts:144]: This should be named `getPresignedUrl` or `presignUrl`

File: packages/@eventual/compiler/src/ast-util.ts [hunk 18]
```diff
@@ -61,6 +61,24 @@ export function isEntityStreamMemberCall(call: CallExpression): boolean {
   return false;
 }
 
+/**
+ * A heuristic for identifying a {@link CallExpression} that is a call to an `on` handler.
+ *
+ * 1. must be a call to a MemberExpression matching to `<expression>.on(events, name, impl | props, impl)`.
+ * 2. must have 3 or 4 arguments.
+ */
+export function iBucketHandlerMemberCall(call: CallExpression): boolean {
```
Comment (sam-goodwin) [inline packages/@eventual/compiler/src/ast-util.ts:70]: `is`?

**Review by sam-goodwin — CHANGES_REQUESTED** [review]

File: apps/tests/aws-runtime/test/test-service.ts (hunk 11 above)
Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts:706]: V3 returns a stream.Readable in node.

The return type from V3 has a transformToString on it.

https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/index.html#streams

File: packages/@eventual/aws-cdk/src/bucket-service.ts (hunk 13 above)
Comment (thantos) [reply packages/@eventual/aws-cdk/src/bucket-service.ts:188]: Pre Signed url needs cors.

File: packages/@eventual/aws-cdk/src/bucket-service.ts [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+    scope: Construct,
+    id: string,
+    props: BucketNotificationHandlerProps
+  ) {
+    super(scope, id);
+
+    const handlerName = props.handler.spec.name;
+    const bucketName = props.handler.spec.bucketName;
+
+    this.handler = new ServiceFunction(this, "Handler", {
```
[the remark this answers is not in the document]
Comment (thantos) [reply packages/@eventual/aws-cdk/src/bucket-service.ts:240]: Dlq is not supported for the s3 notifications. We could roll our own.

File: packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts (hunk 15 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts:41]: Same thing it does in cdk. A getter for a value. Added when we moved the env variable assertions to be lazy instead of water.

File: packages/@eventual/compiler/src/ast-util.ts (hunk 18 above)
Comment (thantos) [reply packages/@eventual/compiler/src/ast-util.ts:70]: Ha

File: packages/@eventual/aws-cdk/src/bucket-service.ts (hunk 12 above)
Comment (thantos) [reply packages/@eventual/aws-cdk/src/bucket-service.ts:159]: It gives direct access to the Bucket via the `service.buckets` member. However, it could be an interface instead of the construct I guess.

File: packages/@eventual/aws-runtime/src/handlers/command-worker.ts (hunk 16 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/handlers/command-worker.ts:27]: I made them explicit undefined instead of being able to omit the property so that we are forced to update all locations that use this object when adding a new dependency. In this case it is undefined because the spec is injected at handler invocation time instead of creation time as it is used with the serviceUrl that is taken from the headers instead of the env variable like the rest of the handlers.

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 17 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts:144]: HMMMMMMMMM, I don't like `getPresignedUrl` as the presigned url need to be given an operation where `getPresignedUrl("get")` and `getPresignedUrl("put")` are kind of odd.

AWS uses `S3RequestPresigner.signUrl` an `getSignedUrl` in the SDK.

How about `generatePresignedUrl`?

File: apps/tests/aws-runtime/test/test-service.ts (hunk 11 above)
Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts:706]:
```ts
/**
 * The interface contains mix-in utility functions to transfer the runtime-specific
 * stream implementation to specified format. Each stream can ONLY be transformed
 * once.
 */
export interface SdkStreamMixin {
    transformToByteArray: () => Promise<Uint8Array>;
    transformToString: (encoding?: string) => Promise<string>;
    transformToWebStream: () => ReadableStream;
}
```

Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts:706]: Lets do....

`get(key).toString(encoding?: string)` which calls `transformToString` and caches the first result.

File: packages/@eventual/aws-cdk/src/bucket-service.ts (hunk 13 above)
Comment (thantos) [reply packages/@eventual/aws-cdk/src/bucket-service.ts:193]: So it was logical to give the same cors that the api has access to. If there are more use cases, we can add overrides later.

File: packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts (hunk 15 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/handlers/bucket-handler-worker.ts:41]: Lazy just means a value that isn't immediately resolved, as decided by the caller, it doesn't imply a promise, future, or a thread as controlled by the sub-process itself.

File: packages/@eventual/aws-runtime/src/stores/bucket-store.ts (hunk 17 above)
Comment (thantos) [reply packages/@eventual/aws-runtime/src/stores/bucket-store.ts]: Updated

File: apps/tests/aws-runtime/test/test-service.ts (hunk 11 above)
Comment (thantos) [reply apps/tests/aws-runtime/test/test-service.ts]: Added

**Review by sam-goodwin — APPROVED** [review]
